tracim/tracim/lib/email_fetcher.py

```diff
@@ -0,0 +1,296 @@
+                return part
+        # check for plain text
+        for part in self._message.walk():
+            ctype = part.get_content_type()
+            cdispo = str(part.get('Content-Disposition'))
+            if ctype == 'text/plain' and 'attachment' not in cdispo:
+                return part
+        return part
+
+    def get_key(self) -> typing.Optional[str]:
+
+        """
+        First try checking special header, them check 'to' header
+        and finally check first(oldest) mail-id of 'references' header
+        """
+        key = None
+        first_ref = self.get_first_ref()
+        to_address = self.get_to_address()
+        special_key = self.get_special_key()
+
+        if special_key:
+            key = special_key
+        if not key and to_address:
+            key = DecodedMail.find_key_from_mail_address(to_address)
+        if not key and first_ref:
+            key = DecodedMail.find_key_from_mail_address(first_ref)
+
+        return key
+
+    @staticmethod
+    def find_key_from_mail_address(mail_address: str) \
+            -> typing.Optional[str]:
+        """ Parse mail_adress-like string
+        to retrieve key.
+
+        :param mail_address: user+key@something like string
+        :return: key
+        """
+        username = mail_address.split('@')[0]
+        username_data = username.split('+')
+        if len(username_data) == 2:
+            key = username_data[1]
+        else:
+            key = None
+        return key
+
+
+class MailFetcher(object):
+
+    def __init__(self,
+                 host: str, port: str, user: str, password: str, folder: str,
+                 delay: int, endpoint: str, token:str) \
+            -> None:
+        """
+        Fetch mail from a mailbox folder through IMAP and add their content to
+        Tracim through http according to mail Headers.
+        Fetch is regular.
+        :param host: imap server hostname
+        :param port: imap connection port
+        :param user: user login of mailbox
+        :param password: user password of mailbox
+        :param folder: mail folder where new mail are fetched
+        :param delay: seconds to wait before fetching new mail again
+        :param endpoint: tracim http endpoint where decoded mail are send.
+        :param token: token to authenticate http connexion
+        """
+        self._connection = None
+        self._mails = []
+        self.host = host
+        self.port = port
+        self.user = user
+        self.password = password
+        self.folder = folder
+        self.delay = delay
+        self.endpoint = endpoint
+        self.token = token
+
+        self._is_active = True
+
+    def run(self) -> None:
+        while self._is_active:
+            time.sleep(self.delay)
+            self._connect()
+            self._fetch()
+            self._notify_tracim()
+            self._disconnect()
+
+    def stop(self) -> None:
+        self._is_active = False
+
+    def _connect(self) -> None:
```

Comment:
you should add a parameter `connection_life_duration` which does the following : 

```
       now = datetime.now()
       expiration_datetime = self._connection_creation_datetime() + connection_life_duration
       if self._connection:
           if now < expiration_datetime:
               return
           self._disconnect()
      [...]
```

This allow not to disconnect if _connect() is called twice in the same time